docs/token-contracts/fa12/4-fa12-archetype.md

```diff
@@ -0,0 +1,447 @@
+
+```archetype
+forall tokenholder in ledger,
+tokenholder.holder <> %from ->
+tokenholder.holder <> %to ->
+before.ledger[tokenholder.holder] = some(tokenholder)
+```
+> Tokenholders other than `%from` and `%to`, are not modified nor added to `ledger`.
+
+##### 1.5/ Removed `ledger` records
+
+```archetype
+removed.ledger.isempty()
+```
+> No record is removed from `ledger`.
+
+##### 1.6/ Added `ledger` records
+
+```archetype
+let some before_to = before.ledger[%to] in
+  added.ledger.isempty()
+otherwise
+  added.ledger = [ { holder = %to; tokens = value } ]
+```
+> The added record in 'ledger', if any, is the `%to` record.
+
+#### 2/ Effect on `allowance`
+
+##### 2.1/ Effect on `(%from,caller)` allowance record
+
+```archetype
+caller <> %from ->
+let some before_from_caller = before.allowance[(%from,caller)] in
+let some after_from_caller = allowance[(%from,caller)] in
+  before_from_caller.amount > value ->
+  after_from_caller = { before_from_caller with
+    amount = (before_from_caller.amount - value)
+  }
+otherwise false
+otherwise true
+```
+
+> When `caller` is different from `%from`, the amount `caller` is authorised to spend on the behalf of `%from` is decreased by `value` if `value` is striclty greated than the authorized amount.
+
+##### 2.1/ No effect on `allowance`
+```archetype
+caller = %from -> allowance = before.allowance
+```
+> No effect on `allowance` when `caller` is equal to `%from`.
+
+##### 2.2/ Unchanged `allowance` records
+```archetype
+forall a in allowance,
+a.addr_owner <> %from and a.addr_spender <> caller ->
+before.allowance[(a.addr_owner,a.addr_spender)] = some(a)
+```
+
+> Allowed amounts other than those associated to `%from` and `caller` are identical.
+
+##### 2.3/ Added and removed `allowance` records
+```archetype
+removed.allowance.isempty() and added.allowance.isempty()
+```
+> No allowance record is added or removed.
+
+#### 3/ Explicit `fail`
+
+##### 3.1/ Not Enough Balance
+
+```archetype
+fails with (msg : string) :
+  let some after_ledger_from = ledger[%from] in
+    msg = "NotEnoughBalance" and
+    after_ledger_from.tokens < value
+  otherwise true
+```
+
+> When the entry fails with message "NotEnoughBalance", `value` is stricly greater than the number of tokens of `%to`.
+> Cannot spend more than you own.
+
+##### 3.2/ Not Enough Allowance
+```archetype
+fails with (msg : string * (nat * nat)) :
+  let some after_allowance_from_caller = allowance[(%from,caller)] in
+    msg = ("NotEnoughAllowance", (value, after_allowance_from_caller.amount)) and
+    caller <> %from and
+    after_allowance_from_caller.amount < value
+  otherwise false
+```
+> When the entry fails with message "NotEnoughAllowance", `caller` is different from `%from` and `value` is stricly greater than the allowed amount for `%from` and `caller`.
+> A spender cannot spend more than allowed.
+
+#### 4/ Effect on `operations`
+```archetype
+length (operations) = 0
+```
+> No operation generated.
+
+### approve
```

Comment:
The section on transfer is already pretty long and gives a good overview/salespitch of Archetype's specification power, I would put the other entrypoints in an annex (maybe just as comments in the original source?) and just refer to that to avoid loosing the reader  (we do want them to get to the why3 calls down there :arrow_down: ).

Reply:
Done: 
* removed `approve` `getAllowance` `getTotalSupply` `getBalance` sections
* added `other entry point properties` with a link to the [doc](https://app.gitbook.com/@edukera/s/archetype/contract-library/tokens/fa-1.2/formal-properties)